### PR TITLE
Add support to new goth versions

### PR DESCRIPTION
**Pull request by alejandrodevs**
Add support for new goth versions.

A couple of functions need to be added to the Waffle.Definition module:

```elixir
  # Specify the Goth module to authenticate in GCP.
  def goth_module do
    MyApp.Goth.Storage
  end

  # Defines the service account file to be used to sign the urls.
  def service_account_path do
    Application.fetch_env!(:my_app, :gcp_storage)
    |> Keyword.get(:service_account_path)
  end
```

### CODE REVIEW

File: lib/waffle/storage/google/util.ex [hunk 1]
```diff
@@ -3,12 +3,6 @@ defmodule Waffle.Storage.Google.Util do
   A collection of utility functions.
   """
 
-  alias GoogleApi.Gax.{Request, Response}
-  alias GoogleApi.Storage.V1.Connection
-  alias GoogleApi.Storage.V1.Model.Object
-
-  @library_version Mix.Project.config() |> Keyword.get(:version, "")
```
Comment (javierg) [inline lib/waffle/storage/google/util.ex:10]: why did you remove this?

File: lib/waffle/storage/google/url_v2.ex [hunk 2]
```diff
@@ -79,18 +79,9 @@ defmodule Waffle.Storage.Google.UrlV2 do
...
-
-    signature = definition
-    |> build_path(path)
-    |> canonical_request(expiration)
-    |> sign_request()
-
-    base_url = build_url(definition, path)
-
-    "#{base_url}?GoogleAccessId=#{client_id}&Expires=#{expiration}&Signature=#{signature}"
+    client = GcsSignedUrl.Client.load_from_file(definition.service_account_path())
```
Comment (javierg) [inline lib/waffle/storage/google/url_v2.ex]: instad of using load from file and have this extra configuration, you can use

Goth.Config.get(:client_email) and use the strategy as shown in: https://github.com/alexandrubagu/gcs_signed_url#example

with oauth config

Comment (javierg) [reply lib/waffle/storage/google/url_v2.ex]: Goth.Config.get allows yuou to retrieve any value from the goth configuration

**Review by javierg — COMMENTED** [review]
IMHO you shoudn't be changing the URL v2, but add url v4 and allow users to select which version they want to use... Or if there is a drop of features and support for v2 gcs API, then push the version for the waffle gcs adapter

File: lib/waffle/storage/google/cloud_storage.ex [hunk 3]
```diff
@@ -71,9 +69,9 @@ defmodule Waffle.Storage.Google.CloudStorage do
   Constructs a new connection object with scoped authentication. If no scope is
   provided, the `devstorage.full_control` scope is used as a default.
   """
-  @spec conn(String.t) :: Tesla.Env.client
-  def conn(scope \\ @full_control_scope) do
-    {:ok, token} = Goth.Token.for_scope(scope)
+  @spec conn(Types.definition) :: Tesla.Env.client
+  def conn(definition) do
+    {:ok, token} = Goth.fetch(definition.goth_module())
```
Comment (javierg) [inline lib/waffle/storage/google/cloud_storage.ex:74]: on eweay to do this without having to pass the module is to use GenServer functions to retrieve the config and get a  token

```
config_process = GenServer.whereis(Goth.Config)
%{default: config} = :sys.get_state(config_process)
{:ok, token} = Goth.Token.fetch(config)
```

Since the tool relies heavily on Goth anyways this should get you what we need.

File: lib/waffle/storage/google/util.ex (hunk 1 above)
Comment (alejandrodevs) [reply lib/waffle/storage/google/util.ex:10]: It is not been used, I just removed the warnings.

File: lib/waffle/storage/google/url_v2.ex (hunk 2 above)
Comment (alejandrodevs) [reply lib/waffle/storage/google/url_v2.ex]: Goth.Config module is deprecated. We shouldn't use it anymore.
https://github.com/peburrows/goth/blob/master/CHANGELOG.md#v130-2022-06-29

File: lib/waffle/storage/google/cloud_storage.ex (hunk 3 above)
Comment (alejandrodevs) [reply lib/waffle/storage/google/cloud_storage.ex:74]: Goth.Config is deprecated.
https://github.com/peburrows/goth/blob/master/CHANGELOG.md#v130-2022-06-29

**Review by javierg — COMMENTED** [review]
Please update the readme fiel to reflect changes in API and configuration requirements

File: README.md [hunk 4]
```diff
@@ -60,13 +60,26 @@ config :waffle,
   storage_dir: "uploads/waffle"
 ```
 
+Also, add the following functions to your definition module:
```
Comment (javierg) [inline README.md:63]: can you add why these are needed?

Comment (javierg) [inline README.md]: I will gowith something like:

```
In order to be compatible with newer Goth directions we will need to define the next modules:
```

**Review by javierg — APPROVED** [review]
:shipit:
